### PR TITLE
Example of folder structure for selection on DeepSportRadar Instance Segmentation Challenge

### PR DESCRIPTION
**Pull request by Ra1Nik**
This is an example of folder structure of a complete run on [DeepSportRadar Instance Segmentation Challenge](https://github.com/DeepSportRadar/instance-segmentation-challenge)!
Predictions are computed on the union of val and test sets using a detectron2 mask_rcnn_R_50_FPN_3x model finetuned on the train set. Metadata are created with the name of the folders in which images are contained.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,2 +1,25 @@
-# object_detection_example_structure
-Example of folder structure of a complete run on DeepSportRadar Instance Segmentation Challenge dataset
+### Object Detection example on a selection on SportRadar Challenge
```
Comment (IgorSusmelj) [inline README.md:1]: Would rename to `Object Detection Example in Lightly Format`

File: README.md [hunk 2]
```diff
@@ -1,2 +1,25 @@
...
+
+This is an example of folder structure of a complete run on [DeepSportRadar Instance Segmentation Challenge](https://github.com/DeepSportRadar/instance-segmentation-challenge)!
```
Comment (IgorSusmelj) [inline README.md]: Would change this slightly to:
```
This is an example of folder structure in the Lightly format for data selection using data from the [DeepSportRadar Instance Segmentation Challenge](https://github.com/DeepSportRadar/instance-segmentation-challenge)!
```

File: lightly_bucket/.lightly/metadata/KS-FR-BLOIS/24330/camcourt1_1513714448590_0.json [hunk 3]
```diff
@@ -0,0 +1 @@
+{"file_name": "KS-FR-BLOIS/24330/camcourt1_1513714448590_0.png", "type": "image", "metadata": {"court": "KS-FR-BLOIS", "game_id": "24330"}}
```
Comment (IgorSusmelj) [inline lightly_bucket/.lightly/metadata/KS-FR-BLOIS/24330/camcourt1_1513714448590_0.json]: When dumping the metadata as json files I suggest using the `indent=4` parameter to format them more nicely.

E.g. 
```python
with open(path_to_prediction, 'w') as f:
    json.dump(lightly_prediction, f, indent=4) 
```

File: lightly_bucket/.lightly/predictions/object_detection_sportradar/schema.json [hunk 4]
```diff
@@ -0,0 +1 @@
+{"task_type": "object-detection", "categories": [{"id": 0, "name": "human", "supercategory": "none"}]}
```
Comment (IgorSusmelj) [inline lightly_bucket/.lightly/predictions/object_detection_sportradar/schema.json]: Also here, try to dump this using `indent=4` for nicer/ more human readable formatting :)

File: worker.sh [hunk 5]
```diff
@@ -0,0 +1,2 @@
+#! /usr/bin/bash
+docker run --shm-size="1024m" --gpus all --rm -it     -v OUTPUT_DIR:/home/output_dir lightly/worker:latest token=your_awesome_token worker.worker_id=your_worker_id
```
Comment (IgorSusmelj) [inline worker.sh]: Let's stick to the convention in the rest of the examples (e.g. https://docs.lightly.ai/docs/install-lightly#register-the-lightly-worker)
- no output volume mounting
- pass token using environment variable
- use `{}` and capital letters for the placeholders
- use linebreak

All together this would look like this:
```bash
docker run --shm-size="1024m" --gpus all --rm -it -e LIGHTLY_TOKEN={MY_LIGHTLY_TOKEN} \ 
    lightly/worker:latest token=your_awesome_token worker.worker_id={MY_WORKER_ID}
```

File: README.md [hunk 6]
```diff
@@ -1,2 +1,25 @@
...
+Then simply run the worker with your API token and your worker id:
+
+```
+bash worker.sh
+```
+
+And finally run your Python scheduling script with:
+
+```
+python job_scheduler.py
```
Comment (IgorSusmelj) [inline README.md]: Can we call this script something along the lines of `run_lightly.py`?
(I'd try to use **Run** instead of **Job**)

File: README.md [hunk 7]
```diff
@@ -1,2 +1,25 @@
...
+
+```
+aws s3 cp input_bucket s3://yourInputBucket/sportradar_input/ --recursive
+aws s3 cp lightly_bucket s3://yourInputBucket/sportradar_lightly/ --recursive
+```
+
+Then simply run the worker with your API token and your worker id:
+
+```
+bash worker.sh
```
Comment (IgorSusmelj) [inline README.md]: Let's maybe call this script `start_lightly_worker.sh`?

File: README.md [hunk 8]
```diff
@@ -1,2 +1,25 @@
-# object_detection_example_structure
-Example of folder structure of a complete run on DeepSportRadar Instance Segmentation Challenge dataset
+### Object Detection example on a selection on SportRadar Challenge
+
+This is an example of folder structure of a complete run on [DeepSportRadar Instance Segmentation Challenge](https://github.com/DeepSportRadar/instance-segmentation-challenge)!
+Predictions are computed on the union of val and test sets using a detectron2 mask_rcnn_R_50_FPN_3x model finetuned on the train set. Metadata are created with the name of the folders in which images are contained.
+
```
Comment (IgorSusmelj) [inline README.md:6]: I'd drop here a rough folder structure as a reference such that users get an idea what to expect. You can use tree.

**Review by IgorSusmelj — CHANGES_REQUESTED** [review]
This looks already very good. I have left a few comments/ suggestions.

Furthermore, please also make use of the relevant filenames feature :)
- e.g. to select a few specific images
- and also to select somewhere a full folder (using `/*`)

**Review by IgorSusmelj — APPROVED** [review]
This looks very good! I'd add a small note/sentence in the beginning that makes it clear that this example uses code to work with AWS but the folder structure for other cloud providers stays the same.

File: README.md [hunk 9]
```diff
@@ -1,2 +1,60 @@
...
+        │       │   └── ...
+        │       └── ...
+        └── relevant_filenames.txt
+```
+
+In order to run the selection (on valtest set) you have first to upload data to your cloud buckets by using the following AWS CLI commands:
+
+```
+aws s3 cp input_bucket s3://yourInputBucket/sportradar_input/ --recursive
+aws s3 cp lightly_bucket s3://yourInputBucket/sportradar_lightly/ --recursive
```
Comment (guarin) [inline README.md]: I would use `aws s3 cp lightly_bucket s3://yourLightlyBucket/sportradar_lightly/ --recursive` instead of `aws s3 cp lightly_bucket s3://yourInputBucket/sportradar_lightly/ --recursive`. This highlights that two different buckets can be used for input and lightly.

File: README.md [hunk 10]
```diff
@@ -1,2 +1,60 @@
...
+```
+
+Then simply run the worker with your API token and your worker id:
```
Comment (guarin) [inline README.md]: I would drop the "simply" here. It might not be that simple for users. As a user I don't like it if docs say that something is simple and then I struggle with it for a while.

File: README.md [hunk 11]
```diff
@@ -1,2 +1,60 @@
...
+
+```
+bash start_lightly_worker.sh
+```
```
Comment (guarin) [inline README.md]: I would copy the docker run command here as it is pretty short.

File: README.md (hunk 10 above)
Comment (guarin) [inline README.md]: Could you also add a link to our docs on how to install the lightly worker?

File: README.md [hunk 12]
```diff
@@ -1,2 +1,60 @@
...
+
+And finally run your Python scheduling script with:
+
```
Comment (guarin) [inline README.md:60]: Could you add a note that this requires the lightly python client and link to the docs on how to install it?

File: start_lightly_worker.sh [hunk 13]
```diff
@@ -0,0 +1,3 @@
+#! /usr/bin/bash
+docker run --shm-size="1024m" --gpus all --rm -it -e LIGHTLY_TOKEN={MY_LIGHTLY_TOKEN} \ 
+    lightly/worker:latest token=your_awesome_token worker.worker_id={MY_WORKER_ID}
```
Comment (guarin) [inline start_lightly_worker.sh:3]: Could you drop the `token=your_awesome_token` part? The token is already passed with `LIGHTLY_TOKEN={MY_LIGHTLY_TOKEN}`.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,60 @@
-# object_detection_example_structure
-Example of folder structure of a complete run on DeepSportRadar Instance Segmentation Challenge dataset
+### Object Detection example on a selection on SportRadar Challenge
+
+This is an example of folder structure in the Lightly format for data selection using data from the [DeepSportRadar Instance Segmentation Challenge](https://github.com/DeepSportRadar/instance-segmentation-challenge)!
+Predictions are computed on the union of val and test sets using a detectron2 mask_rcnn_R_50_FPN_3x model finetuned on the train set. Metadata are created with the name of the folders in which images are contained. KS-FR-ROANNE folder is removed using our [relevant filenames feature](https://docs.lightly.ai/docs/relevant-filenames).
```
Comment (guarin) [inline README.md:4]: I didn't find any mention of `KS-FR-ROANNE` in the relevant filenames file.

**Review by guarin — COMMENTED** [review]
This looks really nice, great work!

I left a couple comments. We should also check the license whether we are allowed to redistribute the images and if we have to include a license ourselves.

File: run_lightly.py [hunk 15]
```diff
@@ -0,0 +1,124 @@
+from lightly.api import ApiWorkflowClient
+from lightly.openapi_generated.swagger_client.models.dataset_type import DatasetType
+from lightly.openapi_generated.swagger_client.models.datasource_purpose import DatasourcePurpose
+
+# Create the Lightly client to connect to the API.
+client = ApiWorkflowClient(token='your_awesome_token')
```
Comment (guarin) [inline run_lightly.py]: `client = ApiWorkflowClient(token="MY_LIGHTLY_TOKEN")`

File: run_lightly.py [hunk 16]
```diff
@@ -0,0 +1,124 @@
...
+
+# Create a new dataset on the Lightly Platform.
+client.create_dataset(
+    'your_dataset_name',
+    DatasetType.IMAGES
+)
+
+# Input bucket
+client.set_s3_config(
+    resource_path='s3://YourInputBucket/sportradar_input/',
```
Comment (guarin) [inline run_lightly.py]: Bucket name is not consistent with the aws command in the readme.

File: run_lightly.py [hunk 17]
```diff
@@ -0,0 +1,124 @@
...
+    region='eu-central-1',
+    access_key='your_access_key',
+    secret_access_key='your_secret_access_key',
+    purpose=DatasourcePurpose.INPUT
+)
+
+# Lightly bucket
+client.set_s3_config(
+    resource_path='s3://YourLightlyBucket/sportradar_lightly/',
```
Comment (guarin) [inline run_lightly.py]: Bucket name is not consistent with the aws command in the readme.

File: run_lightly.py (hunk 15 above)
Comment (guarin) [reply run_lightly.py]: In general, please follow the same format and naming as in our docs: https://docs.lightly.ai/docs/set-up-your-first-dataset

File: run_lightly.py [hunk 18]
```diff
@@ -0,0 +1,124 @@
...
+    region='eu-central-1',
+    access_key='your_access_key',
+    secret_access_key='your_secret_access_key',
+    purpose=DatasourcePurpose.LIGHTLY
+)
+
+client.schedule_compute_worker_run(
+    worker_config={
+        'enable_corruptness_check': True,
+        'remove_exact_duplicates': True,
```
Comment (guarin) [inline run_lightly.py]: Please drop `enable_corruptness_check ` and `remove_exact_duplicates `. They are True by default and want to move users away from using those options.

File: run_lightly.py [hunk 19]
```diff
@@ -0,0 +1,124 @@
...
+            }
+        }
+        ]
+    },
+    lightly_config={
+        'loader': {
+            'batch_size': 16,
+            'shuffle': True,
+            'num_workers': -1,
+            'drop_last': True
```
Comment (guarin) [inline run_lightly.py]: Not sure if we need the full lightly config here. Maybe only the changed arguments are enough.

File: README.md (hunk 14 above)
Comment (Ra1Nik) [reply README.md:4]: I canceled the KS-FR-ROANNE from the relevant filename list, which comprehends every filename but the KS-FR-ROANNE ones. Isn't it the proper way to do that?
